ADBDEV-3884: Fix planning failure on modifying CTE with gp_cte_sharing enabled.

KnightMurloc opened the pull request:
When the planner was preparing shared scan over CTE, that contained modifying
operation, the Materialize node did not retrieved the target list from
ModifyTable node. That led to the init of Materialize node was failed with
the error: "Material operator does not support projection", because of
that the number of elements in its target list and the child node are different.
It also led to Shared Scan node also doesn't had target list. This happened
because ModifyTable node don't set its own target list on creation.

This patch sets target list of Materialize node on it's creation from returning
lists of child ModifyTable node. We pick first returning list because the first
one (in the case of inherited tables) contains the target list of the root table
while the rest of the child tables. We can take the returning list before it
processing in set_plan_refs because Materialize node has its own processing
in set_plan_refs. Returning list is processed in set_returning_clause_references
which handles situations with outer Vars. Since ModifyTable does the projection,
we don't need to handle these situations.

KnightMurloc committed:
Fix incorrect plan when using CTE with gp_cte_sharing enabled.

The problem is that when creating a T_ModifyTable node, the targetlist is not specified for it, and therefore the targetlist is empty in the parent nodes T_Material and T_ShareInputScan, which leads to an error.

Fixed by specifying targetlist after creating T_ModifyTable node in subquery_planner function. targetlist is retrieved from the returningList of the Query structure. Also removing the Assert that the targetlist is empty and filling targetlist it in the set_plan_references function since it is now specified at an earlier stage.

RekGRpth commented on src/backend/optimizer/plan/planner.c:
```diff
@@ -908,6 +908,14 @@ subquery_planner(PlannerGlobal *glob, Query *parse,
+			 * targetlist. Also the targetlist might be needed by higher-level nodes,
+			 * such as Material, for correct operation.
+			 */
+			plan->targetlist = copyObject(parse->returningList);
```
The make_modifytable function has a returningLists argument, which is passed parse->returningList. Why not do both changes in the same place in the make_modifytable function?

KnightMurloc replied:
make_modifytable takes a list of lists, and it's not clear which one should be used.

RekGRpth replied:
Solution is partial revert https://github.com/arenadata/gpdb/commit/9fa82c980935ef4aee18fabe8da20ae2198b052a but why?!
```diff
diff --git a/src/backend/optimizer/plan/createplan.c b/src/backend/optimizer/plan/createplan.c
index 986b27889b..5b453ca2c0 100644
--- a/src/backend/optimizer/plan/createplan.c
+++ b/src/backend/optimizer/plan/createplan.c
@@ -6367,8 +6367,10 @@ make_modifytable(PlannerInfo *root,
 	node->plan.lefttree = NULL;
 	node->plan.righttree = NULL;
 	node->plan.qual = NIL;
-	/* setrefs.c will fill in the targetlist, if needed */
-	node->plan.targetlist = NIL;
+	if (returningLists)
+		node->plan.targetlist = copyObject(linitial(returningLists));
+	else
+		node->plan.targetlist = NIL;
 
 	node->operation = operation;
 	node->canSetTag = canSetTag;
diff --git a/src/backend/optimizer/plan/setrefs.c b/src/backend/optimizer/plan/setrefs.c
index 98eefd64e8..39f7f95208 100644
--- a/src/backend/optimizer/plan/setrefs.c
+++ b/src/backend/optimizer/plan/setrefs.c
@@ -1044,7 +1044,6 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
 			{
 				ModifyTable *splan = (ModifyTable *) plan;
 
-				Assert(splan->plan.targetlist == NIL);
 				Assert(splan->plan.qual == NIL);
 
 				splan->withCheckOptionLists =
@@ -1088,7 +1087,6 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
 					 * we don't have to do set_returning_clause_references()
 					 * twice on identical targetlists.
 					 */
-					splan->plan.targetlist = copyObject(linitial(newRL));
 				}
 
 				foreach(l, splan->resultRelations)
```

RekGRpth commented:
On 7X problem was solved by another way, but resulting plan is same!

KnightMurloc commented:
I checked on 7X a different plan is generated and execution fails.

<details>
<summary>6X with patch</summary>

```
postgres=# explain verbose with cte1 as (insert into t1 values (1,2) returning *) select * from cte1 a join cte1 b using(c1);
                                                            QUERY PLAN                                                            
----------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice4; segments: 3)  (cost=0.05..0.13 rows=4 width=440)
   Output: share0_ref2.c1, share0_ref2.c2, b.c2
   ->  Hash Join  (cost=0.05..0.13 rows=2 width=440)
         Output: share0_ref2.c1, share0_ref2.c2, b.c2
         Hash Cond: (share0_ref2.c1 = b.c1)
         ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..0.04 rows=1 width=222)
               Output: share0_ref2.c1, share0_ref2.c2
               Hash Key: share0_ref2.c1
               ->  Shared Scan (share slice:id 1:0)  (cost=0.01..0.22 rows=1 width=0)
                     Output: share0_ref2.c1, share0_ref2.c2
         ->  Hash  (cost=0.04..0.04 rows=1 width=222)
               Output: b.c2, b.c1
               ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=0.00..0.04 rows=1 width=222)
                     Output: b.c2, b.c1
                     Hash Key: b.c1
                     ->  Subquery Scan on b  (cost=0.01..0.23 rows=1 width=0)
                           Output: b.c2, b.c1
                           ->  Shared Scan (share slice:id 3:0)  (cost=0.01..0.22 rows=1 width=0)
                                 Output: share0_ref1.c1, share0_ref1.c2
                                 ->  Materialize  (cost=0.00..0.01 rows=1 width=0)
                                       Output: c1, c2
                                       ->  Insert on public.t1  (cost=0.00..0.01 rows=1 width=0)
                                             Output: c1, c2
                                             ->  Redistribute Motion 1:3  (slice2; segments: 1)  (cost=0.00..0.01 rows=1 width=0)
                                                   Output: 1, '2'::character varying(100)
                                                   ->  Result  (cost=0.00..0.01 rows=1 width=0)
                                                         Output: 1, '2'::character varying(100)
 Optimizer: Postgres query optimizer
 Settings: gp_cte_sharing=on
```

</details>

<details>
<summary>7X</summary>

```
postgres=# explain verbose with cte1 as (insert into t1 values (1,2) returning *) select * from cte1 a join cte1 b using(c1);
                                                          QUERY PLAN                                                          
------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.16..0.34 rows=3 width=440)
   Output: a.c1, a.c2, b.c2
   ->  Hash Join  (cost=0.16..0.30 rows=1 width=440)
         Output: a.c1, a.c2, b.c2
         Hash Cond: (a.c1 = b.c1)
         ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..0.12 rows=3 width=222)
               Output: a.c1, a.c2
               Hash Key: a.c1
               ->  Subquery Scan on a  (cost=0.00..0.06 rows=3 width=222)
                     Output: a.c1, a.c2
                     ->  Shared Scan (share slice:id 2:0)  (cost=0.03..0.23 rows=1 width=222)
         ->  Hash  (cost=0.12..0.12 rows=3 width=222)
               Output: b.c2, b.c1
               ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=0.00..0.12 rows=3 width=222)
                     Output: b.c2, b.c1
                     Hash Key: b.c1
                     ->  Subquery Scan on b  (cost=0.00..0.06 rows=3 width=222)
                           Output: b.c2, b.c1
                           ->  Shared Scan (share slice:id 3:0)  (cost=0.03..0.23 rows=1 width=222)
                                 ->  Insert on public.t1  (cost=0.00..0.03 rows=1 width=222)
                                       Output: t1.c1, t1.c2
                                       ->  Redistribute Motion 1:3  (slice4; segments: 1)  (cost=0.00..0.03 rows=1 width=222)
                                             Output: 1, '2'::character varying(100)
                                             ->  Result  (cost=0.00..0.01 rows=1 width=222)
                                                   Output: 1, '2'::character varying(100)
 Optimizer: Postgres query optimizer
 Settings: gp_cte_sharing = 'on'
```


</details>

As you can see, the Materialise node has disappeared in 7X. most likely that's why the planning does not fall. but SharedScan nodes still don't have output. therefore, the execution of the plan still fails.

RekGRpth commented:
OK, how about suggested changes?

KnightMurloc commented:
It is proposed to take the first returningList from returningLists in make_modifytable as it was before https://github.com/arenadata/gpdb/commit/9fa82c980935ef4aee18fabe8da20ae2198b052a ?

KnightMurloc commented:
might it be worth revert [9fa82c9](https://github.com/arenadata/gpdb/commit/9fa82c980935ef4aee18fabe8da20ae2198b052a) completely, leaving only the tests?

RekGRpth commented:
> the Materialise node has disappeared in 7X



> SharedScan nodes still don't have output. therefore, the execution of the plan still fails

Yes, you are right, there is a problem on 7x too.

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -6367,8 +6367,19 @@ make_modifytable(PlannerInfo *root,
-	/* setrefs.c will fill in the targetlist, if needed */
-	node->plan.targetlist = NIL;
+
+	if(returningLists){
```
Bad formatting! In original it was
```c
	/*
	 * Set up the visible plan targetlist as being the same as the first
	 * RETURNING list.	This is for the use of EXPLAIN; the executor won't pay
	 * any attention to the targetlist.
	 */
	if (returningLists)
		node->plan.targetlist = copyObject(linitial(returningLists));
	else
		node->plan.targetlist = NIL;
```

RekGRpth commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1079,16 +1078,6 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
-					 * we don't have to do set_returning_clause_references()
-					 * twice on identical targetlists.
-					 */
-					splan->plan.targetlist = copyObject(linitial(newRL));
```
Why don't you use the `set_returning_clause_references` function for `plan.targetlist `anymore?

KnightMurloc replied:
I think it probably makes sense to copy the `returningList` after `set_returning_clause_references` since it is used later.

KnightMurloc committed:
added copying of returningList to targetlist after set_returning_clause_references.
fixed formatting.

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -6367,8 +6367,18 @@ make_modifytable(PlannerInfo *root,
+	 * targetlist. Also the targetlist might be needed by higher-level nodes,
+	 * such as Material, for correct operation.
+	 */
+	if(returningLists)
```
missed space between if and (

RekGRpth commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1081,12 +1080,7 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
-					 * twice on identical targetlists.
+					 * copy the new returningList as it may be needed later.
 					 */
 					splan->plan.targetlist = copyObject(linitial(newRL));
```
Maybe you should free the list first?

KnightMurloc replied:
can't other nodes refer to it? it seems to be deleted anyway when the context is deleted.

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -6367,8 +6367,18 @@ make_modifytable(PlannerInfo *root,
-	node->plan.targetlist = NIL;
+
+	/*
+	 * Set up the visible plan targetlist as being the same as
```
Why didn't you return the old comment exactly as it was?

KnightMurloc committed:
fix comment

RekGRpth commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1081,12 +1080,7 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
-					 * twice on identical targetlists.
+					 * copy the new returningList as it may be needed later.
 					 */
 					splan->plan.targetlist = copyObject(linitial(newRL));
```
Maybe in this place find empty target lists for the material and shared scan nodes and fill them in?

KnightMurloc replied:
are we can travel through plan tree from bottom to up?

RekGRpth replied:
Or, at a slightly later stage, go down from the specified nodes with empty target lists and fill them.

KnightMurloc replied:
this approach does not fix the problem since one of the Shared Scans does not have bottom nodes. look like it get a targetlist at an earlier stage.

bimboterminator1 commented on src/backend/optimizer/plan/setrefs.c:
```diff
@@ -1081,12 +1080,7 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
-					 * twice on identical targetlists.
+					 * copy the new returningList as it may be needed later.
 					 */
 					splan->plan.targetlist = copyObject(linitial(newRL));
```
> can't other nodes refer to it?

In other words, Material node uses the original returningList, that had been copied to ModifyTable targetList inside make_modifytable. Then,  inside the set_plan_refs,  ModifyTable replaces its targetList and returningList by a newRL. From that point Material and ModifyTable may have different targetLists. Can it be potentially be harmful?

KnightMurloc replied:
As I understand it, the `set_returning_clause_references` function only changes `varna` and this comment says that this is a normal situation that they may differ in Material and outer plan. https://github.com/arenadata/gpdb/blob/23babbbd5abfe7098a00e7ee72231a0aeef257d0/src/backend/executor/nodeMaterial.c#L352-L358

bimboterminator1 replied:
Yes, execution is not affected. But for EXPLAIN output can there be a problem? When `set_returning_clause_references` changes the ModifyTable's tlist entries, and Material's does not?

KnightMurloc replied:
EXPLAIN uses only attribute names. which should not change.

bimboterminator1 commented on src/test/regress/sql/with.sql:
```diff
@@ -1046,3 +1046,14 @@ WITH updated AS (
+    c1 int,
+    c2 int)
+distributed randomly;
+explain (costs off) with cte1 as (insert into t1 values (1,2) returning *) select * from cte1 a join cte1 b using(c1);
```
What happens for UPDATE/DELETE? The issue and the error remain the same without patch?

bimboterminator1 replied:
Yes, UPDATE/DELETE behave the same way. No need to appen these test cases.

bimboterminator1 commented on src/test/regress/sql/with.sql:
```diff
@@ -1046,3 +1046,14 @@ WITH updated AS (
 select count(*) from rank_tbl where rank in (select rank from updated);
+
+-- checking that the planner can build a plan with non-select CTE sharing.
+set gp_cte_sharing to on;
+drop table if exists t1;
+create table t1 (
+    c1 int,
+    c2 int)
+distributed randomly;
+explain (costs off) with cte1 as (insert into t1 values (1,2) returning *) select * from cte1 a join cte1 b using(c1);
```
1. Wrap `drop table if exists` with:
```
--start_ignore
drop table if exists t1;
--end_ignore
```
2. Write create table query in single line.
3. Write explain query in multi line style:
```
explain (costs off)
with cte1 as (
 insert into t1 values (1,2)
 returning *)
select * from cte1 a join cte1 b using(c1);
```
All these changes help tests to be more readable.

bimboterminator1 replied:
.out should be changed respectively

KnightMurloc committed:
fix test formatting

KnightMurloc committed:
fix test formatting2

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/52635

bimboterminator1 commented:
@KnightMurloc  At version 
```
select version ();
PostgreSQL 9.4.26 (Greenplum Database 6.25.1+dev.573.geb8f68895e build dev) on x86_64-unknown-linux-gnu, compiled by x86_64-alt-linux-gcc (GCC) 10.3.1 20210703 (ALT Sisyphus 10.3.1-alt2), 64-bit compiled on Aug 28 2023 13:00:55 (with ass
ert checking)
(1 row)
```
everything works fine:
```
set gp_cte_sharing to on;
--start_ignore
drop table if exists t1;
--end_ignore
create table t1 (c1 int,c2 int) distributed randomly;
explain (costs off)
with cte1 as (
  insert into t1 values (1,2) returning *
) select * from cte1 a join cte1 b using(c1);
                                          QUERY PLAN                                           
------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice4; segments: 3)
   ->  Hash Join
         Hash Cond: (share0_ref2.c1 = b.c1)
         ->  Redistribute Motion 3:3  (slice1; segments: 3)
               Hash Key: share0_ref2.c1
               ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Redistribute Motion 3:3  (slice3; segments: 3)
                     Hash Key: b.c1
                     ->  Subquery Scan on b
                           ->  Shared Scan (share slice:id 3:0)
                                 ->  Materialize
                                       ->  Insert on t1
                                             ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                                   ->  Result
 Optimizer: Postgres query optimizer
(16 rows)

with cte1 as (
  insert into t1 values (1,2) returning *
) select * from cte1 a join cte1 b using(c1);
 c1 | c2 | c2 
----+----+----
  1 |  2 |  2
(1 row)

select * from t1;
 c1 | c2 
----+----
  1 |  2
(1 row)
```
**Why it works correctly?**

KnightMurloc commented:
yes the execution should also work. do I need to add a test case for it?

RekGRpth commented on src/test/regress/sql/with.sql:
```diff
@@ -1046,3 +1046,16 @@ WITH updated AS (
 select count(*) from rank_tbl where rank in (select rank from updated);
+
+-- checking that the planner can build a plan with non-select CTE sharing.
+set gp_cte_sharing to on;
```
I think that this line is not needed here, because this parameter is set at the beginning of the file; also, it does not need to be restored at the end.

KnightMurloc committed:
fix tests

bimboterminator1 commented:
> yes the execution should also work. do I need to add a test case for it?

Yes, you should add it, please. Because the query is executed correctly and we can show it.

KnightMurloc commented:
should I leave the explain test too, or should I only do the execution test?

bimboterminator1 commented:
> should I leave the explain test too, or should I only do the execution test?

Explain should be there as well.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53053

KnightMurloc committed:
add execute test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53102

bimboterminator1 commented:
@KnightMurloc , @RekGRpth What about forming targetList only for the case of plan sharing? Something like [that](https://github.com/arenadata/gpdb/pull/606/commits/39b9892fae425f119d8b2675e8141b7309492268). At least there won't be targetList creation every time ModifyTable node is created.

KnightMurloc commented:
can there be a situation when we have Material above Modify Table but without SharedScan?

KnightMurloc commented:
i think it's more safe solution in term of side effects but i don't know how good idea fix only specific case.

RekGRpth commented:
> What about forming targetList only for the case of plan sharing? Something like [that](https://github.com/arenadata/gpdb/pull/606/commits/39b9892fae425f119d8b2675e8141b7309492268). At least there won't be targetList creation every time ModifyTable node is created. And there won't be partial revert of legacy patch.

Yes, I also like this idea.

KnightMurloc commented:
OK. let's use this solution.

bimboterminator1 commented:
> can there be a situation when we have Material above Modify Table but without SharedScan?

We should change the solution only if there is no any other similar situation

RekGRpth commented:
> > can there be a situation when we have Material above Modify Table but without SharedScan?
> 
> We should change the solution only if there is no any other similar situation. So, if we proof that, we may consider only current specific case.

The original problem was in non-select CTE sharing and you solution fixes it.

KnightMurloc commented:
as I understood, in addition to Sharedscan, Material only used by joins, and in this case there will be another Subquery Scan node between Material and Modify Table, and in this case the problem is not reproduced.

KnightMurloc committed:
roll back previous changes. And if the Material node does not have a target list
and there is a ModifyTable under it, take the first returning list as the target list.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53263

bimboterminator1 commented on src/backend/optimizer/plan/planshare.c:
```diff
@@ -211,6 +211,14 @@ prepare_plan_for_sharing(PlannerInfo *root, Plan *common)
+			ModifyTable *mt = (ModifyTable *)common;
+			shared->targetlist = copyObject(linitial(mt->returningLists));
+		}
+		Assert(shared->targetlist);
```
Do we need that assert? If Material's targetList will be empty we will get the same error as in current ticket. It's unnecessary, I guess

KnightMurloc replied:
does this assert describe the problem better? because the original one says that the planner requires projections for the node that does not support it, whereas the problem is not in this.

bimboterminator1 commented on src/test/regress/sql/with.sql:
```diff
@@ -1046,3 +1046,20 @@ WITH updated AS (
 )
 select count(*) from rank_tbl where rank in (select rank from updated);
+
+-- checking that the planner can build a plan with non-select CTE sharing.
```
Please, write 'Test' instead 'checking'.

RekGRpth commented on src/backend/optimizer/plan/planshare.c:
```diff
@@ -211,6 +211,14 @@ prepare_plan_for_sharing(PlannerInfo *root, Plan *common)
+			 IsA(common, ModifyTable))
+		{
+			ModifyTable *mt = (ModifyTable *)common;
+			shared->targetlist = copyObject(linitial(mt->returningLists));
```
may be add check?
```c
if (mt->returningLists)
```
or
```c
Assert(mt->returningLists);
```

bimboterminator1 replied:
If we have modifying CTE, parser won't allow to write something without returning list.
```
explain (costs off)
with cte1 as (
  insert into t1 values (1,2)            
) select * from cte1 a join cte1 b using(c1);
ERROR:  WITH query "cte1" does not have a RETURNING clause
LINE 4: ) select * from cte1 a join cte1 b using(c1);
```
Moreover, if we don't have returningList this solution will be equivalent to the initial solution  with NIL targetList

KnightMurloc replied:
`linitial` and `copyObject` normally handle NULL values. so no verification is required.

andr-sokolov replied:
Fully expanded linitial is `((list_head(mt->returningLists))->data.ptr_value)`. The list_head returns NULL, when its argument is NULL. So if mt->returningLists is NULL, then segfault happens. I suggest to add a check

KnightMurloc committed:
fix test comment

RekGRpth commented on src/backend/optimizer/plan/planshare.c:
```diff
@@ -211,6 +211,14 @@ prepare_plan_for_sharing(PlannerInfo *root, Plan *common)
 		Material *m = make_material(common);
 		shared = (Plan *) m;
 
+		if (!shared->targetlist &&
```
These conditions do not correlate?
If common is ModifyTable then may shared->targetlist be not null?
May be leave only check if common is ModifyTable?

KnightMurloc committed:
remove assert and useless check

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/53298

bimboterminator1 commented:
@KnightMurloc, Rewrite the PR and commit description in accordance with current changes please. Also extend the description of the problematic behaviour.

andr-sokolov commented:
@KnightMurloc , Does the patch work correctly when CTE is update on a table which has inherited tables? Why is it correct to get the first element of mt->returningLists as targetlist when mt->returningLists has more then one element?

KnightMurloc commented:
several `returningLists` appear in case of update or delete on the parent table, then `returningLists` contains, in addition to the `returningList` of parent table, also a `returningList` of child tables, so it makes sense to take the first `returningList`.

KnightMurloc committed:
change test to uppercase

KnightMurloc committed:
fix test formating

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54031

andr-sokolov commented on src/test/regress/sql/with.sql:
```diff
@@ -1046,3 +1046,22 @@ WITH updated AS (
+WITH cte1 AS (
+	INSERT INTO t1 VALUES ( 1, 2 ) RETURNING *
+)
+SELECT * FROM cte1 a JOIN cte1 b USING(c1);
```
Add please space after USING in this line too.

KnightMurloc committed:
fix test formating

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54099

andr-sokolov commented on src/test/regress/sql/with.sql:
```diff
@@ -1046,3 +1046,22 @@ WITH updated AS (
+--end_ignore
+CREATE TABLE t1 (c1 int, c2 int) DISTRIBUTED RANDOMLY;
+
+EXPLAIN (COSTS OFF)
```
Add VERBOSE to show target lists

KnightMurloc committed:
add verbose to test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54110

KnightMurloc committed:
fix formating

KnightMurloc committed:
add check that mt->returningLists not null

andr-sokolov commented:
What about patching `make_material` instead of `prepare_plan_for_sharing`?

```
diff --git a/src/backend/optimizer/plan/createplan.c b/src/backend/optimizer/plan/createplan.c
index 986b27889b..fe03c8df7c 100644
--- a/src/backend/optimizer/plan/createplan.c
+++ b/src/backend/optimizer/plan/createplan.c
@@ -5751,7 +5751,15 @@ make_material(Plan *lefttree)
        Plan       *plan = &node->plan;
 
        /* cost should be inserted by caller */
-       plan->targetlist = lefttree->targetlist;
+       if (lefttree->targetlist != NIL)
+               plan->targetlist = lefttree->targetlist;
+       else if (IsA(lefttree, ModifyTable))
+       {
+               ModifyTable *mt = (ModifyTable *)lefttree;
+               Assert(mt->returningLists);
+               plan->targetlist = copyObject(linitial(mt->returningLists));
+       }
+
        plan->qual = NIL;
        plan->lefttree = lefttree;
        plan->righttree = NULL;

```

KnightMurloc committed:
setup targetlist for Material node in case of ModifyTable lefttree

RekGRpth commented on src/backend/optimizer/plan/createplan.c:
```diff
@@ -5751,7 +5751,15 @@ make_material(Plan *lefttree)
 
 	/* cost should be inserted by caller */
-	plan->targetlist = lefttree->targetlist;
+	if(lefttree->targetlist != NIL)
```
pgindent your changes

RekGRpth commented:
What will be the final commit message?

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/54267

bimboterminator1 commented:
@KnightMurloc Please, rewrite the patch discription, add more info about the problem and motivation behind the final solution. And about solution itself.
> The problem was reproduced when 

Description should briefly but succinctly tell what the problem was about, the context of the issue. You can start it like: "When the planner was preparing shared scan over CTE, that contained modifying operation, the Materialize node did not retrieved the target list from ModifyTable node. That led to the ...."
>target list at that stage

What stage is meant exactly? The description should allow the reviewer to recognize the issue at first glance.

bimboterminator1 commented:
> because Materialize node has its own processing in set_plan_refs.

Write more about the causes why it's safe to take the returning list before it processing in set_plan_ref. How set_plan_refs influences the tlists and ensure the reviewer that current changes are safe.

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-3884
